zephyr/wrapper.c

```diff
@@ -68,9 +68,9 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
 
 #else
 
-extern uint8_t _end, _heap_sentry;
-#define heapmem ((uint8_t *)ALIGN_UP((uintptr_t)&_end, PLATFORM_DCACHE_ALIGN))
-#define HEAPMEM_SIZE (&_heap_sentry - heapmem)
+extern char _end[], _heap_sentry[];
+#define heapmem ((uint8_t *)ALIGN_UP((uintptr_t)_end, PLATFORM_DCACHE_ALIGN))
+#define HEAPMEM_SIZE ((uint8_t *)_heap_sentry - heapmem)
```

Comment:
interesting, https://sourceware.org/binutils/docs-2.38/ld.html#Assignments specifies referencing linker defined symbols closer to what SOF is doing now...

Reply:
From https://sourceware.org/binutils/docs-2.38/ld.html#Source-Code-Reference

> Accessing a linker script defined variable from source code is not intuitive. 
 [...] Alternatively the symbols can be treated as the names of vectors or arrays and then the code will again work as expected

Reply:
Linker script variables are not intuitive.

Array to pointer decay is not intuitive http://c-faq.com/aryptr/aryptrequiv.html

I find it really nice to use both so they cancel each other out.

Also, it seems very accurate to type a random memory location as an incomplete array of bytes. This makes sure very little can be done with it without casting it first.

Reply:
@marc-hb nice, thanks, didn't notice that "alternative" in the document